Source/WebCore/Modules/url-pattern/URLPattern.cpp

```diff
@@ -26,28 +26,243 @@
 #include "config.h"
 #include "URLPattern.h"
 
+#include "URLPatternCanonical.h"
 #include "URLPatternInit.h"
 #include "URLPatternOptions.h"
 #include "URLPatternResult.h"
 #include <wtf/RefCounted.h>
+#include <wtf/URL.h>
 
 namespace WebCore {
 
 WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(URLPattern);
 
-Ref<URLPattern> URLPattern::create(URLPatternInput&&, String&& baseURL, URLPatternOptions&&)
+template<typename CharacterType>
+static String escapePatternStringForCharacters(std::span<const CharacterType> characters)
+{
+    static constexpr std::array escapeCharacters { '+', '*', '?', ':', '(', ')', '\\', '{', '}' }; // NOLINT
+
+    StringBuilder result;
+    result.reserveCapacity(characters.size());
+
+    for (auto character : characters) {
+        if (std::find(escapeCharacters.begin(), escapeCharacters.end(), character) != escapeCharacters.end())
+            result.append('\\');
+
+        result.append(character);
+    }
+
+    return result.toString();
+}
+
+// https://urlpattern.spec.whatwg.org/#escape-a-pattern-string
+static String escapePatternString(StringView input)
+{
+    ASSERT(input.containsOnlyASCII());
+
+    if (input.is8Bit())
+        return escapePatternStringForCharacters(input.span8());
+
+    return escapePatternStringForCharacters(input.span16());
+}
+
+// https://urlpattern.spec.whatwg.org/#process-a-base-url-string
+static String processBaseURLString(StringView input, BaseURLStringType type)
+{
+    if (type != BaseURLStringType::Pattern)
+        return input.toString();
+
+    return escapePatternString(input);
+}
+
+// https://urlpattern.spec.whatwg.org/#process-a-urlpatterninit
+static ExceptionOr<URLPatternInit> processInit(URLPatternInit&& init, BaseURLStringType type, String&& protocol = String(), String&& username = String(), String&& password = String(), String&& hostname = String(), String&& port = String(), String&& pathname = String(), String&& search = String(), String&& hash = String())
+{
+    URLPatternInit result;
+    result.protocol = WTFMove(protocol);
+    result.username = WTFMove(username);
+    result.password = WTFMove(password);
+    result.hostname = WTFMove(hostname);
+    result.port = WTFMove(port);
+    result.pathname = WTFMove(pathname);
+    result.search = WTFMove(search);
+    result.hash = WTFMove(hash);
+
+    URL baseURL;
+
+    if  (!init.baseURL.isNull()) {
+        baseURL = URL(init.baseURL);
+
+        if (!baseURL.isValid() || baseURL.isEmpty())
```

Comment:
Are we sure empty baseURL is not allowed? Do we have a test for this?

Reply:
> Do we have a test for this?

Tests are not enabled yet

Reply:
The subsequent code is using baseURL.user(), baseURL.password(), etc to populate the URLPattern. 

URLParser accepts the empty string as valid input so I guess I thought it made sense because if the baseURL string is empty, these fields should be null-opt anyways. Which is no change from the initial initializations. 

eg - port() is not going to be different (aka is null anyways) if we process an empty baseURL string. 
<img width="786" alt="Screenshot 2024-10-23 at 10 09 05 AM" src="https://github.com/user-attachments/assets/30c2bc01-d42e-4faf-968b-f852bbbd6f13">

Reply:
I understand what you said, but you are throwing exception on empty URL here, which means it is an invalid parameter. From your description, it's a valid parameter, it's just we are not going to update some fields of result based on `baseURL`, right?

Reply:
I guess my thought was it's better to throw a failure rather than to have some side-effects like updating Protocol to an empty String instead of a default null String, etc.